### PR TITLE
New Skeletal Resource aws_s3_buckets

### PR DESCRIPTION
**Pull request by dromazmj**
This PR adds a new resource for testing all or a group of s3 buckets

### CODE REVIEW

**Review by jquick — APPROVED** [review]
Thanks for this @dromazmj ! This one will also need rebased.

File: docs/resources/aws_s3_buckets.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+---
+title: About the aws_s3_buckets Resource
+---
+
+# aws_s3_buckets
```
Comment (clintoncwolfe) [inline docs/resources/aws_s3_buckets.md]: We need to escape all underscores in titles, etc - anywhere they are unquoted.  Thanks!

File: docs/resources/aws_s3_buckets.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+Use the `aws_s3_buckets` InSpec audit resource to list all buckets in a single account.
```
Comment (clintoncwolfe) [inline docs/resources/aws_s3_buckets.md]: You might also refer to the `aws_s3_bucket` resource to perform in-depth auditing of an individual bucket.

File: docs/resources/aws_s3_buckets.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+The control will pass if the resource contains at least one bucket.
+
+    # Test if there are any buckets
+    describe aws_s3_buckets
+      it { should exist }
+    end
+
+## Properties
+
+### bucket_names
```
Comment (clintoncwolfe) [inline docs/resources/aws_s3_buckets.md]: :100: for using plural property names on a plural resource!

File: docs/resources/aws_s3_buckets.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Provides an array of strings containing the names of the buckets.
+
+    # Examine a specific subnets cidr_blocks
+    describe aws_s3_buckets do
+      its('bucket_names') { should eq ['my_bucket'] }
```
Comment (clintoncwolfe) [inline docs/resources/aws_s3_buckets.md]: Might also include an `include` test (which won't care about array length or order):

`its('bucket_names') { should include 'my_bucket' }`

File: lib/resources/aws/aws_s3_buckets.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
+# author: Matthew Dromazos
+# author: Sam Cornwell
+class AwsS3Buckets < Inspec.resource(1)
+  name 'aws_s3_buckets'
+  desc 'Verifies settings for AWS S3 Buckets in bulk'
+  example "
+    describe aws_s3_bucket do
+      its('Bucket_names') { should eq ['my_bucket'] }
```
Comment (clintoncwolfe) [inline lib/resources/aws/aws_s3_buckets.rb]: That `Bucket_names` should probably be `bucket_names`.

File: lib/resources/aws/aws_s3_buckets.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  include AwsPluralResourceMixin
+
+  # Underlying FilterTable implementation.
+  filter = FilterTable.create
+  filter.add_accessor(:where)
+        .add_accessor(:entries)
+        .add(:exists?) { |x| !x.entries.empty? }
+        .add(:bucket_names, field: :name)
+        .add(:creation_dates, field: :creation_date)
```
Comment (clintoncwolfe) [inline lib/resources/aws/aws_s3_buckets.rb]: Looks like this isn't documented.  Unless that was intentional, it should be added to the docs.

File: test/aws/default/verify/controls/aws_s3_buckets.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  # You should be able to test if there are any buckets
+  describe aws_s3_buckets do
+    it { should exist }
+  end
+end
+
+control "aws_s3_buckets properties" do
+  # you should be able to test the cidr_block of a subnet
+  describe aws_s3_buckets do
+    its('bucket_names') { should include fixtures['s3_bucket_public_name'], fixtures['s3_bucket_private_name'] }
```
Comment (clintoncwolfe) [inline test/aws/default/verify/controls/aws_s3_buckets.rb]: Could you add a negative test here?  Perhaps

`its('bucket_names') { should_not include 'i-dont-exist' }`

File: test/unit/resources/aws_s3_buckets_test.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+#=============================================================================#
+#                               Test Fixtures
+#=============================================================================#
+module AwsMSBB
+  class Basic < AwsBackendBase
+    def list_buckets
+      fixtures = [
+        OpenStruct.new({
+          name: "bucket_01",
```
Comment (clintoncwolfe) [inline test/unit/resources/aws_s3_buckets_test.rb]: If you're going to include creation date, be sure to include it here.

File: test/unit/resources/aws_s3_buckets_test.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+#=============================================================================#
+module AwsMSBB
+  class Basic < AwsBackendBase
+    def list_buckets
+      fixtures = [
+        OpenStruct.new({
+          name: "bucket_01",          
+        }),
+        OpenStruct.new({
+          name: "bucket_02",
```
Comment (clintoncwolfe) [inline test/unit/resources/aws_s3_buckets_test.rb]: Also - s3 bucket names must be DNS compliant - so you can use dashes, but not underscores.

File: test/unit/resources/aws_s3_buckets_test.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    assert(basic.bucket_names.include?('bucket_01'))
+    refute(basic.bucket_names.include?(nil))
+  end
+end
+
+#=============================================================================#
+#                               Test Fixtures
+#=============================================================================#
+module AwsMSBB
+  class Basic < AwsBackendBase
```
Comment (clintoncwolfe) [inline test/unit/resources/aws_s3_buckets_test.rb]: You should also define a test fixture class that is empty, so we can test the negative side of `exist`s.

File: test/unit/resources/aws_s3_buckets_test.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+  def test_constructor_no_args_ok
+    AwsS3Buckets.new
+  end
+
+  def test_constructor_reject_unknown_resource_params
+    assert_raises(ArgumentError) { AwsS3Buckets.new(bla: 'blabla') }
+  end
+end
+
```
Comment (clintoncwolfe) [inline test/unit/resources/aws_s3_buckets_test.rb:22]: Here is where you could add a 'Recall' section, where you test `exist` against an empty fixture and populated one.

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Great work again @dromazmj !  Just a few docs tweaks, and it would be great to have a negative unit test for `exist`.  Looking forward to pulling this in!

**Comment by clintoncwolfe** [comment]
Unit and integration tests passed - thanks!

File: test/unit/resources/aws_s3_buckets_test.rb (hunk 10 above)
Comment (dromazmj) [reply test/unit/resources/aws_s3_buckets_test.rb]: Good call!

**Comment by dromazmj** [comment]
@clintoncwolfe good catch on the creation date.  When I was making this resource I thought that would be a good property to have but for now I think its best to just get the resource through and if there is a use case for creation date I can add it.  Thanks!

**Review by clintoncwolfe — APPROVED** [review]
Great work!
